File: sdk/src/auctionSubscriber/auctionSubscriber.ts
import { AuctionSubscriberConfig, AuctionSubscriberEvents } from './types';
import { VelocityClient } from '../velocityClient';
import { getUserFilter, getUserWithAuctionFilter } from '../memcmp';
import StrictEventEmitter from 'strict-event-emitter-types';
import { EventEmitter } from 'events';
import { UserAccount } from '../types';
import { ConfirmOptions, Context, PublicKey } from '@solana/web3.js';
import { WebSocketProgramAccountSubscriber } from '../accounts/webSocketProgramAccountSubscriber';
import { ResubOpts } from '../accounts/types';

export class AuctionSubscriber {
	private velocityClient: VelocityClient;
	private opts: ConfirmOptions;
	private resubOpts?: ResubOpts;

	eventEmitter: StrictEventEmitter<EventEmitter, AuctionSubscriberEvents>;
	private subscriber?: WebSocketProgramAccountSubscriber<UserAccount>;

	constructor({
		velocityClient,
		opts,
		resubTimeoutMs,
		logResubMessages,
	}: AuctionSubscriberConfig) {
		// Type-system guarantees at least one of the two is supplied.
		this.velocityClient = velocityClient!;
		this.opts = opts || this.velocityClient.opts;
		this.eventEmitter = new EventEmitter();
		this.resubOpts = { resubTimeoutMs, logResubMessages };
	}

	public async subscribe() {
		let subscriber = this.subscriber;
		if (!subscriber) {
			subscriber = new WebSocketProgramAccountSubscriber<UserAccount>(
				'AuctionSubscriber',
				'user',
				this.velocityClient.program,
				(
					this.velocityClient.program.account as any
				).user.coder.accounts.decode.bind(
					(this.velocityClient.program.account as any).user.coder.accounts
				),
				{
					filters: [getUserFilter(), getUserWithAuctionFilter()],
					commitment: this.opts.commitment,
				},
				this.resubOpts
			);
			this.subscriber = subscriber;
		}

		await subscriber.subscribe(
			(accountId: PublicKey, data: UserAccount, context: Context) => {
				this.eventEmitter.emit(
					'onAccountUpdate',
					data,
					accountId,
					context.slot
				);
			}
		);
	}

	public async unsubscribe() {
		if (!this.subscriber) {
			return;
		}
		this.subscriber.unsubscribe();
	}
}

File: rust/drift-rs/crates/src/auction_subscriber.rs
use std::sync::Mutex;

use crate::solana_sdk::commitment_config::CommitmentConfig;
use solana_account_decoder_client_types::UiAccountEncoding;

use crate::{
    drift_idl::accounts::User,
    memcmp::{get_user_filter, get_user_with_auction_filter},
    types::SdkResult,
    websocket_program_account_subscriber::{
        ProgramAccountUpdate, WebsocketProgramAccountOptions, WebsocketProgramAccountSubscriber,
    },
    SdkError, UnsubHandle,
};

pub struct AuctionSubscriberConfig {
    pub commitment: CommitmentConfig,
    pub resub_timeout_ms: Option<u64>,
    pub url: String,
}

/// Subscribes to all user auction events across all markets
///
/// DEV: take care it is not dropped or the Auction stream will unsubscribe
pub struct AuctionSubscriber {
    subscriber: WebsocketProgramAccountSubscriber,
    unsub: Mutex<Option<UnsubHandle>>,
}

impl AuctionSubscriber {
    pub const SUBSCRIPTION_ID: &'static str = "auction";

    pub fn new(config: AuctionSubscriberConfig) -> Self {
        let filters = vec![get_user_filter(), get_user_with_auction_filter()];
        let websocket_options = WebsocketProgramAccountOptions {
            filters,
            commitment: config.commitment,
            encoding: UiAccountEncoding::Base64Zstd,
        };

        Self {
            subscriber: WebsocketProgramAccountSubscriber::new(config.url, websocket_options),
            unsub: Mutex::new(None),
        }
    }

    /// Start the auction subscription task
    ///
    /// * `handler_fn` - fn to invoke on each update
    ///
    /// this class sends the entire User account, the callback is required to
    /// interpret the diff e.g orders added/removed
    ///
    pub fn subscribe<F>(&self, handler_fn: F)
    where
        F: 'static + Send + Fn(&ProgramAccountUpdate<User>),
    {
        let mut guard = self.unsub.try_lock().expect("uncontested");
        let unsub = self.subscriber.subscribe(Self::SUBSCRIPTION_ID, handler_fn);
        guard.replace(unsub);
    }

    /// Unsubscribe stopping the auction subscription task
    pub fn unsubscribe(self) -> SdkResult<()> {
        let mut guard = self.unsub.lock().expect("acquired");
        if let Some(unsub) = guard.take() {
            if unsub.send(()).is_err() {
                log::error!("unsub failed");
                return Err(SdkError::CouldntUnsubscribe);
            }
        }

        Ok(())
    }
}

#[cfg(feature = "rpc_tests")]
mod tests {
    use super::*;
    use crate::utils::test_envs::mainnet_endpoint;

    #[tokio::test]
    async fn test_auction_subscriber() {
        env_logger::init();

        let config = AuctionSubscriberConfig {
            commitment: CommitmentConfig::confirmed(),
            resub_timeout_ms: None,
            url: mainnet_endpoint(),
        };

        let mut auction_subscriber = AuctionSubscriber::new(config);

        let emitter = auction_subscriber.event_emitter.clone();

        emitter.subscribe(move |event| {
            log::info!("{:?}", event.now.elapsed());
        });

        let _ = auction_subscriber.subscribe().await;

        tokio::time::sleep(tokio::time::Duration::from_secs(60)).await;

        let _ = auction_subscriber.unsubscribe().await;

        tokio::time::sleep(tokio::time::Duration::from_secs(5)).await;
    }
}
